Split visualizations into packs

axelboc opened the pull request:
Split out NeXus visualizations from core visualizations by introducing the concept of "visualization packs".

![image](https://user-images.githubusercontent.com/2936402/105368325-b36e3900-5c01-11eb-8785-7714902593d8.png)

`App` now renders a component called `VisPackChooser`, which fetches the entity at the selected path and determines which visualization pack to render. If the entity has a `default` attribute or is an `NXdata` group, then `NexusPack` is rendered; otherwise `CorePack` is rendered.

`CorePack` and `NexusPack` can then determine which of their visualizations are supported by the entity they receive, in which ever way they see fit. They then render the `Visualizer` component, which receives an `activeVis` definition object (with the `Icon`, `Container`, etc. to use).

axelboc reviewed (commented):
I feel like this was worthwhile:

- The NeXus visualization are now more clearly separate from the core visualization in the tree and decoupled from them in the code.
- Letting each pack check which of its visualizations are supported by a given entity is very flexible, and will allow us to more easily fetch entities from the store lazily if need be.
- Almost no duplication was introduced thanks to the sharing of `Visualizer` and the new `useActiveVis` custom hook. We've actually removed duplication by combining the two Nexus `supportsEntity` functions directly into the Nexus pack's `getSupportedVis` function.

axelboc commented:
You should probably pull the branch to get a better feel for the new structure. I'm not sure about the location of a couple of files, like `VisPackChooser`.

loichuder reviewed (changes requested):
I would put `VisPackChooser` and `visualizer/hooks` at the root of `vis-packs`.

Also, it feels weird to have `models` files in a `shared` folder and one in the root of the pack folder. What about having the `shared` folder only contain components while other files (`models`, `utils` and `hooks`) are moved at the root of the pack folder ?

`Visualizer` name is no longer fitting in my opinion. It can stay like this for now but we should perhaps think about another name (we could revert to `VisDisplay` ?)

axelboc commented on src/h5web/vis-packs/nexus/shared/utils.test.ts:
```diff
@@ -108,17 +107,6 @@ describe('NeXus utilities', () => {
-      expect(isNxInterpretation(42)).toBe(false);
-      expect(isNxInterpretation('unknown')).toBe(false);
-    });
-  });
```
`isNxInterpretation` is no longer an exported utility -- it's used by the Nexus packs' `getSupportedVis` utility and is therefore tested by this utility's tests.

axelboc commented on src/h5web/vis-packs/nexus/utils.test.ts:
```diff
@@ -0,0 +1,88 @@
+      const nxDataSignal2D = makeNxDataGroup('foo', { signal: datasetInt2D });
+      const supportedVis = getSupportedVis(nxDataSignal2D);
+
+      expect(supportedVis).toEqual(NEXUS_VIS.image);
```
I'm just as surprise as you, but this is a valid way to access a value in a `Record` indexed by an enum. It autocompletes and everything!

axelboc commented on src/h5web/visualizer/VisPackChooser.tsx:
```diff
@@ -0,0 +1,27 @@
+    return <NexusPack entity={entity} />;
+  }
+
+  return <CorePack entity={entity} />;
```
`VisPackChooser`? The logic is covered by the E2E specs. I don't think integration tests would add any value.

axelboc commented:
> I would put VisPackChooser and visualizer/hooks at the root of vis-packs.

Yeah, I wasn't sure because you could argue that the `Visualizer` itself is also a shared component of the two packs. Since we extracted it into its own root folder, I thought maybe the other things that were shared by the two packs had to go in the same folder, but no, they're different things, you're right. 👍 

> Also, it feels weird to have models files in a shared folder and one in the root of the pack folder. What about having the shared folder only contain components while other files (models, utils and hooks) are moved at the root of the pack folder ?

Yeah, the problem was differentiating the pack's utilities (`getSupportedVis`, `getDefaultPath`), from the visualization's utilities (which are numerous). Perhaps we could rename the files `pack-utils` and `vis-utils` (or keep the latter as `utils`)? 

> Visualizer name is no longer fitting in my opinion. It can stay like this for now but we should perhaps think about another name (we could revert to VisDisplay ?)

True. It's a shame that `vis-display` wouldn't be a great name for the folder...

loichuder commented:
> Yeah, the problem was differentiating the pack's utilities (`getSupportedVis`, `getDefaultPath`), from the visualization's utilities (which are numerous). Perhaps we could rename the files `pack-utils` and `vis-utils` (or keep the latter as `utils`)?

`pack-utils` and `utils` seems good to me.

axelboc commented:
Cool, moving the utility/model/hook files out of the `shared` folders really cleans things up. 💯 

- I was able to merge two `hooks` files in the core pack (`core/hooks` which contained `useDatasetValue/s` and `core/shared/hooks`).
- I moved the raw and scalar vis and their CSS files in sub-folders to match the other core visualisations.
- I moved the Spectrum vis toolbar config from `nexus/shared/config` to `nexus/spectrum/config`. It's the only file in the folder, but it matches the structure inside the core pack.

loichuder reviewed (approved):
Looks good to me !
